eclipse-theia/theia
# Improve Preference Filtering

Signed-off-by: Colin Grant <colin.grant@ericsson.com>



#### What it does


This PR aims to make the filtering of nodes in the preference widget more performant, and to smooth the way for feature enhancements like #8252.

At the moment, filtering and scope changing actually modify the `root` of the TreeModel, rather than maintaining the same root and conditionally rendering tree nodes based on various criteria. The performance fix in #7936 improves the situation, but at the cost of some flexibility in scoping preferences. This PR moves that functionality into a modified `TreeModel` implementation and uses an iterative strategy parallel to the `doUpdateRows` function in the `TreeWidget` to determine which nodes to display. This will make filtering more performant as well as more flexible in response to changing demands.

In addition, this PR refactors away some of the not-well-encapsulated code from #7105 by eliminating the `PreferenceEventService` and locating emitters and state tracking in the classes logically responsible for them.

Also fixes #8689

#### How to test


It should retain the functionality of #7105, #7773, #7764, and #7936.

In addition:
- Switching from User/Workspace to Folder scope or vice versa should be faster.
1. Open a multi-root or saved workspace.
1. Open preferences widget.
1. Switch between User scope and Folder scope.
1. Observe that the transition is quick, with little visible delay.

- Filter performance is improved.
1. Open preference widget
1. Type a text query in the search bar.
1. Observe that the filter is smooth, with little delay.

- Node visibility is preserved better.
1. Open preferences widget.
1. Scroll in editor until some preference in the section `css.lint` is at the top of the visible window.
1. Switch to folder scope.
1. Observe that the same preference is still visible. (Formerly this would have caused a scroll to top.)
1. Type the name of the preference in the filter bar.
1. Observe that the preference is still at the top of the screen. (Formerly a filter would have caused a scroll to top.)

- User and Workspace scopes are distinguished properly
1. Follow the steps outlined in #8689 
2. Observe that the transition between User and Workspace scopes is managed correctly.

Further:
- Filter behavior can now be customized by modifying or overriding the `passesCurrentFilters()` method of the `PreferenceTreeModel`. This will mean that implementing enhancements like #8252 is straightforward and performant.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
